vyper/semantics/types/indexable/sequence.py

```diff
@@ -208,6 +218,41 @@ def from_annotation(
         )
 
 
+class DynamicArrayFunctionDefinition(BaseTypeDefinition):
+    """
+    Dynamic array function type definition.
+
+    This class has no corresponding primitive.
+    """
+
+    _is_callable = True
+
+    def __init__(self, name: str, min_arg_count: int, max_arg_count: int) -> None:
+        super().__init__(DataLocation.UNSET)
+        self.name = name
+        self.min_arg_count = min_arg_count
+        self.max_arg_count = max_arg_count
+
+    def __repr__(self):
+        return f"dynamic array function '{self.name}'"
+
+    def fetch_call_return(self, node: vy_ast.Call) -> Optional[BaseTypeDefinition]:
+        validate_call_args(node, (self.min_arg_count, self.max_arg_count))
+
+        dynarray_definition = validation.utils.get_exact_type_from_node(node.func.value)
+        value_type = dynarray_definition.value_type
+        dynarray_length = dynarray_definition.length
+
+        if self.name == "append":
+            validation.utils.validate_expected_type(node.args[0], value_type)
+            return DynamicArrayDefinition(value_type, dynarray_length + 1)
```

Comment:
two things -- I think `append()` should return void, and also it won't change the length bound on the dynarray, it will only change the runtime length.